feat: resource tags in dataset

keunsoopark opened the pull request:
- [x] Make sure to open an issue as a [bug/issue](https://github.com/googleapis/python-bigquery/issues/new/choose) before writing your code!  That way we can discuss the change, evaluate designs, and agree on the general idea
- [x] Ensure the tests and linter pass
- [x] Code coverage does not decrease (if any source code was changed)
- [ ] Appropriate docs were updated (if necessary)

Fixes https://github.com/googleapis/python-bigquery/issues/2091 🦕

Commit:
feat: resource tags in dataset

Commit:
fix: fix unittets

Linchin committed:
Delete dataset/pyvenv.cfg

Linchin commented on google/cloud/bigquery/dataset.py:
```diff
@@ -801,6 +802,26 @@ def labels(self, value):
+    def resource_tags(self):
+        """Dict[str, str]: Resource tags of the dataset.
+
+        This method always returns a dict. To change a dataset's resource tags,
+        modify the dict, then call
+        :meth:`google.cloud.bigquery.client.Client.update_dataset`. To delete
+        a resource tag, set its value to :data:`None` before updating.
```
Change the docstring to be consistent with [the API doc](https://cloud.google.com/bigquery/docs/reference/rest/v2/datasets).
```suggestion
        Optional. The tags attached to this dataset. Tag keys are globally
        unique. Tag key is expected to be in the namespaced format, for
        example "123456789012/environment" where 123456789012 is
        the ID of the parent organization or project resource for this tag
        key. Tag value is expected to be the short name, for example
        "Production".
```

Linchin commented on google/cloud/bigquery/dataset.py:
```diff
@@ -801,6 +802,26 @@ def labels(self, value):
+
+    @resource_tags.setter
+    def resource_tags(self, value):
+        if not isinstance(value, dict):
```
```suggestion
        if not isinstance(value, dict) or value is not None:
```

Linchin reviewed (changes requested):
Thank you so much for adding resource tags in dataset! The PR looks great overall, I just made some minor modifications. I wonder if you could add a system test, as well as a unit test for setter with None? Please let us know if you need any help. Happy holidays!

keunsoopark committed:
Update google/cloud/bigquery/dataset.py

Co-authored-by: Lingqing Gan <lingqing.gan@gmail.com>

keunsoopark committed:
Update google/cloud/bigquery/dataset.py

Co-authored-by: Lingqing Gan <lingqing.gan@gmail.com>

Commit:
added system tests & fix unittest for none

Commit:
add missing assert

Commit:
remove venv

keunsoopark commented:
> system test

Hi @Linchin, thanks for the review.

I added system tests in [system/test_clients.py](https://github.com/keunsoopark/python-bigquery/blob/3d317d1b0a182df7d7f9707e243a4081dcd783de/tests/system/test_client.py#L301)

I fixed unittests for None, and added more tests for invalid values. See [test_resource_tags_setter_bad_value](https://github.com/keunsoopark/python-bigquery/blob/3d317d1b0a182df7d7f9707e243a4081dcd783de/tests/unit/test_dataset.py#L908).

Merry Christmas and happy new year!

Linchin committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Linchin committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Linchin committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Linchin commented:
I think we also need to install the resource-manager package, similar to [this change](https://github.com/googleapis/python-bigquery/pull/2093/files#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835). Could you add this code too?

Commit:
include resourcemanager in noxfile.py

keunsoopark committed:
Merge branch 'main' into feat/resource_tags_in_dataset

keunsoopark commented:
> I think we also need to install the resource-manager package, similar to [this change](https://github.com/googleapis/python-bigquery/pull/2093/files#diff-f7a16a65f061822bcc73b8296f4dc837353d379d8d9cc5307982cb6941442835). Could you add this code too?

Thanks :) Updated!

Commit:
fix fixture for tag keys

Commit:
register tags before using in tests

Linchin commented on tests/system/test_client.py:
```diff
@@ -278,33 +286,72 @@ def test_create_dataset_with_default_rounding_mode(self):
         self.assertTrue(_dataset_exists(dataset))
         self.assertEqual(dataset.default_rounding_mode, "ROUND_HALF_EVEN")
 
+    def _create_resource_tag_key_and_values(self, key, values):
```
In general you can just use pytest to run system tests, it's slightly easier than nox, and should be able to run on any python version:
```
pytest tests/system/test_client.py
```
You can specify the test class or method too:
```
pytest tests/system/test_client.py::TestBigQuery::test_update_dataset
```
This should make it faster.

Linchin replied:
I think you also need to run lint, that will need to use nox:
```
nox -r -s lint
```

keunsoopark replied:
Thanks I ran linting locally.

I still could not run system tests locally. I think I need to specify `GOOGLE_APPLICATION_CREDENTIALS` env var (see [here](https://github.com/googleapis/python-bigquery/blob/a2bebb95c5ef32ac7c7cbe19c3e7a9412cbee60d/noxfile.py#L201)), but not sure what credentials I can use.
I tried to use the credentials of a service account in my personal Google Cloud project, which is the owner of this project, but I got all `PERMISSION_DENIED` errors for dataset creation, table creation etc.

Linchin replied:
I think if you run pytest instead of nox, you don't have to have `GOOGLE_APPLICATION_CREDENTIALS` set up, because we are bypassing `noxfile.py`.

keunsoopark replied:
Ok I managed running system test locally. This is how:
- Make a python script, `install_dependencies.py` to install all dependencies in `noxfile.py` in my local pip.
- For some reason, `google-cloud-testutils` was not installed in this manner, as specified [here](https://github.com/googleapis/python-bigquery/issues/1758#issuecomment-1898893183), so install this one manually.
- Configure my personal GCP project
- Run pytest

keunsoopark committed:
Merge branch 'main' into feat/resource_tags_in_dataset

keunsoopark committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Commit:
handle alreadyexist error

keunsoopark commented on tests/system/test_client.py:
```diff
@@ -278,33 +287,78 @@ def test_create_dataset_with_default_rounding_mode(self):
+        )
+        try:
+            tag_key = tag_key_client.create_tag_key(tag_key=new_tag_key).result()
+        except AlreadyExists:   # When system tests runs in parallel for multiple Python versions, the tag key may already exist
```
Since we run multiple system tests at the same time for different Python versions, we got `AlreadyExists` error for tag key only in one of two system tests.

Linchin replied:
Great, thank you for fixing this!

Linchin committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Commit:
fix: tag keys & values creation & deletion

Commit:
fix comment

Linchin commented:
The 3.12 system test is still failing - after trying to create the resource tag and finding that it already exists, the code tries to access it and gets PermissionDenied error, which indicates that the tag either doesn't exist, or the client doesn't have the permission to access it. I suspect there is a race condition where the tag was created by sys test 3.8 when 3.12 was trying to create it, and deleted when 3.12 tries to access it 😂 Let me rerun the tests to see if it's flaky.

Commit:
make tag keys unique

keunsoopark commented:
> The 3.12 system test is still failing - after trying to create the resource tag and finding that it already exists, the code tries to access it and gets PermissionDenied error, which indicates that the tag either doesn't exist, or the client doesn't have the permission to access it. I suspect there is a race condition where the tag was created by sys test 3.8 when 3.12 was trying to create it, and deleted when 3.12 tries to access it 😂 Let me rerun the tests to see if it's flaky, so we can verify the assumption.
> 
> If this is indeed the case, we may need to append the python version to the tags to avoid conflict.

Thanks for a good inspiration! I agree that we should make each of test running independent. Instead of postfixing python version, I postfixed random characters to tag keys. I see it works well from my local testing.

In previous run, all tests failed all of sudden, and I could no see logs any longer. Could you check if it happens again in next CI tests?

Linchin commented:
Thank you for the prompt reply! Adding a prefix lgtm. I think there was something wrong with the test infra, I hope it doesn't happen again.

Commit:
remove unused import

Linchin committed:
Merge branch 'main' into feat/resource_tags_in_dataset

Linchin reviewed (approved):
Thank you for your contribution, now all tests have passed! 🎉

LGTM.

keunsoopark commented:
> Thank you for your contribution, now all tests have passed! 🎉
> 
> LGTM.

Thank you so much! It was a pleasant review process with you!
My motivation for this contribution is to making dbt able to attach policy tags - https://github.com/dbt-labs/dbt-adapters/issues/577. As you see in this issue, it requires this functionality on tables as well.
I am looking forward to other PR - https://github.com/googleapis/python-bigquery/pull/2093 done soon, but if it takes too long time to get response, I can contribute it as well. Now I can fix this feature pretty quickly :)

Linchin commented:
It was a pleasure working with you! @keunsoopark 

Let's first see if the original author of #2093 responses, then we can decide what to do next.
